README.md
# `co2mon`

[![Actions Status]][github actions] [![Latest Version]][crates.io] [![API docs]][docs.rs]

[Actions Status]: https://github.com/lnicola/co2mon/workflows/ci/badge.svg
[github actions]: https://github.com/lnicola/co2mon/actions
[Latest Version]: https://img.shields.io/crates/v/co2mon.svg
[crates.io]: https://crates.io/crates/co2mon
[API docs]: https://docs.rs/co2mon/badge.svg
[docs.rs]: https://docs.rs/co2mon/

A driver for the Holtek CO₂ USB monitors, tested using a
[TFA-Dostmann AIRCO2NTROL MINI][AIRCO2NTROL MINI] sensor.

[AIRCO2NTROL MINI]: https://www.tfa-dostmann.de/en/produkt/co2-monitor-airco2ntrol-mini/

## Permissions

On Linux, you need to be able to access the USB HID device. For that, you
can save the following `udev` rule to `/etc/udev/rules.d/60-co2mon.rules`:

```text
ACTION=="add|change", SUBSYSTEMS=="usb", ATTRS{idVendor}=="04d9", ATTRS{idProduct}=="a052", MODE:="0666"
```

Then reload the rules and trigger them:

```shell
udevadm control --reload
udevadm trigger
```

Note that the `udev` rule above makes the device accessible to every local user.

## Quick start

```shell
cargo run --example watch
```

## Releases

Release notes are available in [CHANGELOG.md](co2mon/CHANGELOG.md).

## Protocol

The USB HID protocol is not documented, but is a superset of [this one][co2meters] and was [reverse-engineered][had] [before][revspace].

The implementation was inspired by [this one][co2mon].

[co2meters]: https://co2meters.com/Documentation/Other/AN_RAD_0301_USB_Communications_Revised8.pdf
[co2mon]: https://github.com/dmage/co2mon/
[had]: https://hackaday.io/project/5301/
[revspace]: https://revspace.nl/CO2MeterHacking

## License

This project is licensed under either of

* Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or
   [http://www.apache.org/licenses/LICENSE-2.0][LICENSE-APACHE])
* MIT license ([LICENSE-MIT](LICENSE-MIT) or
   [http://opensource.org/licenses/MIT][LICENSE-MIT])

at your option.

[LICENSE-APACHE]: http://www.apache.org/licenses/LICENSE-2.0
[LICENSE-MIT]: http://opensource.org/licenses/MIT


co2mon/CHANGELOG.md
# Changelog

## [2.1.1] - 2021-10-02

### Changed

 - fixed rendering of Linux commands in the docs (@senden9, [#7](https://github.com/lnicola/co2mon/pull/7))
 - marked `Error` as `#[non_exhaustive]`

## [2.1.0] - 2021-04-16

### Changed

 - added support for newer AIRCO2NTROL devices that don't encrypt the data (@elmarx, [#6](https://github.com/lnicola/co2mon/pull/6))
 - marked `Error` as `#[non_exhaustive]`

## [2.0.3] - 2019-12-19

### Fixed

- Fixed the `hidapi` linking `Cargo.toml` features (@tommilligan, [#3](https://github.com/lnicola/co2mon/pull/3))

## [2.0.2] - 2019-11-27

### Fixed

- Fixed panic when displaying timeout errors

## [2.0.1] - 2019-10-20

### Fixed

- Fixed `html_root_url` value

## [2.0.0] - 2019-08-01

### Added

- Implemented `source` for `Error`
- `Sensor` can now be sent across threads
- Reduced size of `Error`

### Changed

- Updated `hidapi` dependency
- Build times are a lot better (`failure` is no longer used)

### Removed

- Removed `serde` feature

## [1.0.0] - 2019-05-17

### Changed

- Simplified timeout logic

## [1.0.0-rc.3] - 2019-05-16

### Added

- Fixed docs.rs build

## [1.0.0-rc.2] - 2019-05-16

### Added

- Improved documentation

## [1.0.0-rc.1] - 2019-05-16

### Added

- Added optional `serde` support
- Added `Sensor::read` to read both values at once
- Improved documentation

### Changed

- `OpenOptions` is no longer consumed when used
- Implemented common traits for `SingleReading` and `OpenOptions`
- Renamed `OpenOptions::with_serial` to `with_serial_number`
- Renamed `Sensor::read` to `read_one`

## [0.1.0-alpha.4] - 2019-02-02

### Fixed

- Added report id to fix the Windows support

## [0.1.0-alpha.3] - 2019-02-02

### Added

- Added read timeout support
- Added features to control `hidapi` linking

### Changed

- Set default read timeout to 5 seconds

### Fixed

- Short USB reads are now reported as errors

## [0.1.0-alpha.2] - 2019-02-01

### Changed

- Bumped `zo-co2` dependency

## [0.1.0-alpha.1] - 2019-02-01

### Added

- Initial release


zg-co2/src/error.rs
use core::fmt::{self, Display, Formatter};

/// A possible error value when decoding a sensor message.
///
/// # Example
///
/// ```
/// let decoded = zg_co2::decode([0x50, 0x04, 0x57, 0xab, 0x00]);
/// assert!(decoded.is_err());
/// ```
#[derive(Debug)]
#[non_exhaustive]
pub enum Error {
    /// The message was invalid (did not finish with `0x0d`).
    InvalidMessage,
    /// The message had a checksum error.
    Checksum,
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        match self {
            Error::InvalidMessage => write!(f, "invalid message"),
            Error::Checksum => write!(f, "checksum error"),
        }
    }
}

#[cfg(feature = "std")]
impl std::error::Error for Error {}

#[cfg(test)]
mod tests {
    #[test]
    fn test_send() {
        fn assert_send<T: Send>() {}
        assert_send::<super::Error>();
    }

    #[test]
    fn test_sync() {
        fn assert_sync<T: Sync>() {}
        assert_sync::<super::Error>();
    }
}


zg-co2/src/serde_types.rs
use super::SingleReading;
use serde::de::{EnumAccess, Error, SeqAccess, Unexpected, VariantAccess, Visitor};
use serde::export;
use serde::ser::SerializeTupleVariant;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::fmt::{self, Formatter};
use std::marker::PhantomData;

impl Serialize for SingleReading {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self {
            SingleReading::Humidity(val) => {
                serializer.serialize_newtype_variant("SingleReading", 0, "Humidity", val)
            }
            SingleReading::Temperature(val) => {
                serializer.serialize_newtype_variant("SingleReading", 1, "Temperature", val)
            }
            SingleReading::CO2(val) => {
                serializer.serialize_newtype_variant("SingleReading", 2, "CO2", val)
            }
            SingleReading::Unknown(kind, val) => {
                let mut tv =
                    serializer.serialize_tuple_variant("SingleReading", 4, "Unknown", 2)?;
                tv.serialize_field(kind)?;
                tv.serialize_field(val)?;
                tv.end()
            }
            _ => unreachable!(),
        }
    }
}

impl<'de> Deserialize<'de> for SingleReading {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        enum Field {
            Humidity,
            Temperature,
            CO2,
            Unknown,
        }

        struct FieldVisitor;

        impl<'de> Visitor<'de> for FieldVisitor {
            type Value = Field;

            fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
                formatter.write_str("variant identifier")
            }

            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                match v {
                    0 => Ok(Field::Humidity),
                    1 => Ok(Field::Temperature),
                    2 => Ok(Field::CO2),
                    3 => Ok(Field::Unknown),
                    _ => Err(Error::invalid_value(
                        Unexpected::Unsigned(v),
                        &"variant index 0 <= i < 4",
                    )),
                }
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: Error,
            {
                match v {
                    "Humidity" => Ok(Field::Humidity),
                    "Temperature" => Ok(Field::Temperature),
                    "CO2" => Ok(Field::CO2),
                    "Unknown" => Ok(Field::Unknown),
                    _ => Err(Error::unknown_variant(v, VARIANTS)),
                }
            }

            fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
            where
                E: Error,
            {
                match v {
                    b"Humidity" => Ok(Field::Humidity),
                    b"Temperature" => Ok(Field::Temperature),
                    b"CO2" => Ok(Field::CO2),
                    b"Unknown" => Ok(Field::Unknown),
                    _ => {
                        let __value = &export::from_utf8_lossy(v);
                        Err(Error::unknown_variant(__value, VARIANTS))
                    }
                }
            }
        }

        impl<'de> Deserialize<'de> for Field {
            #[inline]
            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
            where
                D: Deserializer<'de>,
            {
                deserializer.deserialize_identifier(FieldVisitor)
            }
        }

        struct SingleReadingVisitor<'de> {
            lifetime: PhantomData<&'de ()>,
        }

        impl<'de> Visitor<'de> for SingleReadingVisitor<'de> {
            type Value = SingleReading;
            fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
                formatter.write_str("enum SingleReading")
            }

            fn visit_enum<E>(self, data: E) -> Result<Self::Value, E::Error>
            where
                E: EnumAccess<'de>,
            {
                match data.variant()? {
                    (Field::Humidity, variant) => {
                        variant.newtype_variant().map(SingleReading::Humidity)
                    }
                    (Field::Temperature, variant) => {
                        variant.newtype_variant().map(SingleReading::Temperature)
                    }
                    (Field::CO2, variant) => variant.newtype_variant().map(SingleReading::CO2),
                    (Field::Unknown, variant) => {
                        struct UnknownVisitor<'de> {
                            lifetime: PhantomData<&'de ()>,
                        }

                        impl<'de> Visitor<'de> for UnknownVisitor<'de> {
                            type Value = SingleReading;

                            fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
                                formatter.write_str("tuple variant SingleReading::Unknown")
                            }

                            #[inline]
                            fn visit_seq<S>(self, mut seq: S) -> Result<Self::Value, S::Error>
                            where
                                S: SeqAccess<'de>,
                            {
                                let kind = seq.next_element()?.ok_or_else(|| {
                                    Error::invalid_length(
                                        0,
                                        &"tuple variant SingleReading::Unknown with 2 elements",
                                    )
                                })?;
                                let value = seq.next_element()?.ok_or_else(|| {
                                    Error::invalid_length(
                                        1,
                                        &"tuple variant SingleReading::Unknown with 2 elements",
                                    )
                                })?;
                                Ok(SingleReading::Unknown(kind, value))
                            }
                        }
                        variant.tuple_variant(
                            2,
                            UnknownVisitor {
                                lifetime: PhantomData,
                            },
                        )
                    }
                }
            }
        }

        const VARIANTS: &'static [&'static str] = &["Humidity", "Temperature", "CO2", "Unknown"];
        deserializer.deserialize_enum(
            "SingleReading",
            VARIANTS,
            SingleReadingVisitor {
                lifetime: PhantomData,
            },
        )
    }
}

#[cfg(test)]
mod tests {
    use super::SingleReading;
    use serde_test::Token;

    #[test]
    fn test_serde() {
        let measurement = SingleReading::Humidity(52.3);
        serde_test::assert_tokens(
            &measurement,
            &[
                Token::NewtypeVariant {
                    name: "SingleReading",
                    variant: "Humidity",
                },
                Token::F32(52.3),
            ],
        );

        let measurement = SingleReading::Temperature(20.0);
        serde_test::assert_tokens(
            &measurement,
            &[
                Token::NewtypeVariant {
                    name: "SingleReading",
                    variant: "Temperature",
                },
                Token::F32(20.0),
            ],
        );

        let measurement = SingleReading::CO2(462);
        serde_test::assert_tokens(
            &measurement,
            &[
                Token::NewtypeVariant {
                    name: "SingleReading",
                    variant: "CO2",
                },
                Token::U16(462),
            ],
        );

        let measurement = SingleReading::Unknown(b'R', 10438);
        serde_test::assert_tokens(
            &measurement,
            &[
                Token::TupleVariant {
                    name: "SingleReading",
                    variant: "Unknown",
                    len: 2,
                },
                Token::U8(b'R'),
                Token::U16(10438),
                Token::TupleVariantEnd,
            ],
        );
    }
}


co2mon/src/lib.rs
#![doc(html_root_url = "https://docs.rs/co2mon/2.0.3")]
#![deny(missing_docs)]

//! A driver for the Holtek ([ZyAura ZG][ZG]) CO₂ USB monitors.
//!
//! The implementation was tested using a
//! [TFA-Dostmann AIRCO2NTROL MINI][AIRCO2NTROL MINI] sensor.
//!
//! [AIRCO2NTROL MINI]: https://www.tfa-dostmann.de/en/produkt/co2-monitor-airco2ntrol-mini/
//! [ZG]: http://www.zyaura.com/products/ZG_module.asp
//!
//! # Example usage
//!
//! ```no_run
//! # use co2mon::{Result, Sensor};
//! # fn main() -> Result<()> {
//! #
//! let sensor = Sensor::open_default()?;
//! let reading = sensor.read_one()?;
//! println!("{:?}", reading);
//! #
//! # Ok(())
//! # }
//! ```
//!
//! # Permissions
//!
//! On Linux, you need to be able to access the USB HID device. For that, you
//! can save the following `udev` rule to `/etc/udev/rules.d/60-co2mon.rules`:
//!
//! ```text
//! ACTION=="add|change", SUBSYSTEMS=="usb", ATTRS{idVendor}=="04d9", ATTRS{idProduct}=="a052", MODE:="0666"
//! ```
//!
//! Then reload the rules and trigger them:
//!
//! ```text
//! ## udevadm control --reload
//! ## udevadm trigger
//! ```
//!
//! Note that the `udev` rule above makes the device accessible to every local user.
//!
//! # References
//!
//! The USB HID protocol is not documented, but was [reverse-engineered][had] [before][revspace].
//!
//! [had]: https://hackaday.io/project/5301/
//! [revspace]: https://revspace.nl/CO2MeterHacking

use hidapi::{HidApi, HidDevice};
use std::convert::TryFrom;
use std::ffi::CString;
use std::result;
use std::time::{Duration, Instant};

pub use error::Error;
pub use zg_co2::SingleReading;

mod error;

/// A specialized [`Result`][std::result::Result] type for the fallible functions.
pub type Result<T> = result::Result<T, Error>;

/// A reading consisting of temperature (in °C) and CO₂ concentration (in ppm) values.
///
/// # Example
///
/// ```no_run
/// # use co2mon::{Result, Sensor};
/// # fn main() -> Result<()> {
/// #
/// let sensor = Sensor::open_default()?;
/// let reading = sensor.read()?;
/// println!("{} °C, {} ppm CO₂", reading.temperature(), reading.co2());
/// #
/// # Ok(())
/// # }
#[derive(Debug, Clone, PartialEq, PartialOrd)]
pub struct Reading {
    temperature: f32,
    co2: u16,
}

impl Reading {
    /// Returns the measured temperature in °C.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{Result, Sensor};
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = Sensor::open_default()?;
    /// let reading = sensor.read()?;
    /// println!("{} °C", reading.temperature());
    /// #
    /// # Ok(())
    /// # }
    pub fn temperature(&self) -> f32 {
        self.temperature
    }

    /// Returns the CO₂ concentration in ppm (parts per million).
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{Result, Sensor};
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = Sensor::open_default()?;
    /// let reading = sensor.read()?;
    /// println!("{} ppm CO₂", reading.co2());
    /// #
    /// # Ok(())
    /// # }
    pub fn co2(&self) -> u16 {
        self.co2
    }
}

/// Sensor driver struct.
///
/// # Example
///
/// ```no_run
/// # use co2mon::{Result, Sensor};
/// # fn main() -> Result<()> {
/// #
/// let sensor = Sensor::open_default()?;
/// let reading = sensor.read_one()?;
/// println!("{:?}", reading);
/// #
/// # Ok(())
/// # }
/// ```
pub struct Sensor {
    device: HidDevice,
    key: [u8; 8],
    timeout: i32,
}

impl Sensor {
    /// Opens the sensor device using the default USB Vendor ID (`0x04d9`) and Product ID (`0xa052`) values.
    ///
    /// When multiple devices are connected, the first one will be used.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{Result, Sensor};
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = Sensor::open_default()?;
    /// let reading = sensor.read_one()?;
    /// println!("{:?}", reading);
    /// #
    /// # Ok(())
    /// # }
    pub fn open_default() -> Result<Self> {
        OpenOptions::new().open()
    }

    fn open(options: &OpenOptions) -> Result<Self> {
        let hidapi = HidApi::new()?;

        const VID: u16 = 0x04d9;
        const PID: u16 = 0xa052;

        let device = match options.path_type {
            DevicePathType::Id => hidapi.open(VID, PID),
            DevicePathType::SerialNumber(ref sn) => hidapi.open_serial(VID, PID, &sn),
            DevicePathType::Path(ref path) => hidapi.open_path(&path),
        }?;

        let key = options.key;

        // fill in the Report Id
        let frame = {
            let mut frame = [0; 9];
            frame[1..9].copy_from_slice(&key);
            frame
        };
        device.send_feature_report(&frame)?;

        let timeout = options
            .timeout
            .map(|timeout| timeout.as_millis())
            .map_or(Ok(-1), i32::try_from)
            .map_err(|_| Error::InvalidTimeout)?;

        let air_control = Self {
            device,
            key,
            timeout,
        };
        Ok(air_control)
    }

    /// Takes a single reading from the sensor.
    ///
    /// # Errors
    ///
    /// An error will be returned on an I/O error or if a message could not be
    /// read or decoded.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{Result, Sensor};
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = Sensor::open_default()?;
    /// let reading = sensor.read_one()?;
    /// println!("{:?}", reading);
    /// #
    /// # Ok(())
    /// # }
    pub fn read_one(&self) -> Result<SingleReading> {
        let mut data = [0; 8];
        if self.device.read_timeout(&mut data, self.timeout)? != 8 {
            return Err(Error::InvalidMessage);
        }

        // if the "magic byte" is present no decryption is necessary. This is the case for AIRCO2NTROL COACH
        // and newer AIRCO2NTROL MINIs in general
        let data = if data[4] == 0x0d {
            data
        } else {
            decrypt(data, self.key)
        };
        let reading = zg_co2::decode([data[0], data[1], data[2], data[3], data[4]])?;
        Ok(reading)
    }

    /// Takes a multiple readings from the sensor until the temperature and
    /// CO₂ concentration are available, and returns both.
    ///
    /// # Errors
    ///
    /// An error will be returned on an I/O error or if a message could not be
    /// read or decoded.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{Result, Sensor};
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = Sensor::open_default()?;
    /// let reading = sensor.read()?;
    /// println!("{} °C, {} ppm CO₂", reading.temperature(), reading.co2());
    /// #
    /// # Ok(())
    /// # }
    pub fn read(&self) -> Result<Reading> {
        let start = Instant::now();
        let mut temperature = None;
        let mut co2 = None;
        loop {
            let reading = self.read_one()?;
            match reading {
                SingleReading::Temperature(val) => temperature = Some(val),
                SingleReading::CO2(val) => co2 = Some(val),
                _ => {}
            }
            if let (Some(temperature), Some(co2)) = (temperature, co2) {
                let reading = Reading { temperature, co2 };
                return Ok(reading);
            }

            if self.timeout != -1 {
                let duration = Instant::now() - start;
                if duration.as_millis() > self.timeout as u128 {
                    return Err(Error::Timeout);
                }
            }
        }
    }
}

fn decrypt(mut data: [u8; 8], key: [u8; 8]) -> [u8; 8] {
    data.swap(0, 2);
    data.swap(1, 4);
    data.swap(3, 7);
    data.swap(5, 6);

    for (r, k) in data.iter_mut().zip(key.iter()) {
        *r ^= k;
    }

    let tmp = data[7] << 5;
    data[7] = data[6] << 5 | data[7] >> 3;
    data[6] = data[5] << 5 | data[6] >> 3;
    data[5] = data[4] << 5 | data[5] >> 3;
    data[4] = data[3] << 5 | data[4] >> 3;
    data[3] = data[2] << 5 | data[3] >> 3;
    data[2] = data[1] << 5 | data[2] >> 3;
    data[1] = data[0] << 5 | data[1] >> 3;
    data[0] = tmp | data[0] >> 3;

    for (r, m) in data.iter_mut().zip(b"Htemp99e".iter()) {
        *r = r.wrapping_sub(m << 4 | m >> 4);
    }

    data
}

#[derive(Debug, Clone)]
enum DevicePathType {
    Id,
    SerialNumber(String),
    Path(CString),
}

/// Sensor open options.
///
/// Opens the first available device with the USB Vendor ID `0x04d9`
/// and Product ID `0xa052`, a `0` encryption key and a 5 seconds timeout.
///
/// Normally there's no need to change the encryption key.
///
/// # Example
///
/// ```no_run
/// # use co2mon::{OpenOptions, Result};
/// # use std::time::Duration;
/// # fn main() -> Result<()> {
/// #
/// let sensor = OpenOptions::new()
///     .timeout(Some(Duration::from_secs(10)))
///     .open()?;
/// #
/// # Ok(())
/// # }
#[derive(Debug, Clone)]
pub struct OpenOptions {
    path_type: DevicePathType,
    key: [u8; 8],
    timeout: Option<Duration>,
}

impl Default for OpenOptions {
    fn default() -> Self {
        Self::new()
    }
}

impl OpenOptions {
    /// Creates a new set of options to be configured.
    ///
    /// The defaults are opening the first connected sensor and a timeout of
    /// 5 seconds.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{OpenOptions, Result};
    /// # use std::time::Duration;
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = OpenOptions::new()
    ///     .timeout(Some(Duration::from_secs(10)))
    ///     .open()?;
    /// #
    /// # Ok(())
    /// # }
    pub fn new() -> Self {
        Self {
            path_type: DevicePathType::Id,
            key: [0; 8],
            timeout: Some(Duration::from_secs(5)),
        }
    }

    /// Sets the serial number of the sensor device to open.
    ///
    /// The serial number appears to be the firmware version.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::OpenOptions;
    /// # use std::error::Error;
    /// # use std::ffi::CString;
    /// # use std::result::Result;
    /// # fn main() -> Result<(), Box<Error>> {
    /// #
    /// let sensor = OpenOptions::new()
    ///     .with_serial_number("1.40")
    ///     .open()?;
    /// #
    /// # Ok(())
    /// # }
    pub fn with_serial_number<S: Into<String>>(&mut self, sn: S) -> &mut Self {
        self.path_type = DevicePathType::SerialNumber(sn.into());
        self
    }

    /// Sets the path to the sensor device to open.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::OpenOptions;
    /// # use std::error::Error;
    /// # use std::ffi::CString;
    /// # use std::result::Result;
    /// # fn main() -> Result<(), Box<Error>> {
    /// #
    /// let sensor = OpenOptions::new()
    ///     .with_path(CString::new("/dev/bus/usb/001/004")?)
    ///     .open()?;
    /// #
    /// # Ok(())
    /// # }
    pub fn with_path(&mut self, path: CString) -> &mut Self {
        self.path_type = DevicePathType::Path(path);
        self
    }

    /// Sets the encryption key.
    ///
    /// The key is used to encrypt the communication with the sensor, but
    /// changing it is probably not very useful.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{OpenOptions, Result};
    /// # use std::ffi::CString;
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = OpenOptions::new()
    ///     .with_key([0x62, 0xea, 0x1d, 0x4f, 0x14, 0xfa, 0xe5, 0x6c])
    ///     .open()?;
    /// #
    /// # Ok(())
    /// # }
    pub fn with_key(&mut self, key: [u8; 8]) -> &mut Self {
        self.key = key;
        self
    }

    /// Sets the read timeout.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{OpenOptions, Result};
    /// # use std::time::Duration;
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = OpenOptions::new()
    ///     .timeout(Some(Duration::from_secs(10)))
    ///     .open()?;
    /// #
    /// # Ok(())
    /// # }
    pub fn timeout(&mut self, timeout: Option<Duration>) -> &mut Self {
        self.timeout = timeout;
        self
    }

    /// Opens the sensor.
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use co2mon::{OpenOptions, Result};
    /// # use std::time::Duration;
    /// # fn main() -> Result<()> {
    /// #
    /// let sensor = OpenOptions::new()
    ///     .timeout(Some(Duration::from_secs(10)))
    ///     .open()?;
    /// #
    /// # Ok(())
    /// # }
    pub fn open(&self) -> Result<Sensor> {
        Sensor::open(self)
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_decrypt() {
        let data = [0x6c, 0xa4, 0xa2, 0xb6, 0x5d, 0x9a, 0x9c, 0x08];
        let key = [0; 8];

        let data = super::decrypt(data, key);
        assert_eq!(data, [0x50, 0x04, 0x57, 0xab, 0x0d, 0x00, 0x00, 0x00]);
    }

    #[test]
    fn test_open_options_send() {
        fn assert_send<T: Send>() {}
        assert_send::<super::Sensor>();
        assert_send::<super::OpenOptions>();
    }

    #[test]
    fn test_open_options_sync() {
        fn assert_sync<T: Sync>() {}
        assert_sync::<super::OpenOptions>();
    }
}
